pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,343 @@
+		return err
+	}
+
+	if err = o.scanForReplicationControllers(pods); err != nil {
+		return err
+	}
+	if err = o.deletePods(pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(o.mapper, false, o.out, "node", o.nodeInfo.Name, "drained")
+	return nil
+}
+
+// getInterestingPods returns all the pods on the node, but it skips DaemonSet
+// pods and mirror pods
+func (o *DrainOptions) getInterestingPods() ([]api.Pod, error) {
+	pods := []api.Pod{}
+	podList, err := o.client.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": o.nodeInfo.Name}.AsSelector())
+	if err != nil {
+		return pods, err
+	}
+	for _, pod := range podList.Items {
+		creator, found := pod.ObjectMeta.Annotations["kubernetes.io/created-by"]
+		if found {
+			var daemon_set bool
+			daemon_set, err = o.isDaemonSet(creator)
+			if err != nil {
+				return pods, err
+			}
+			if daemon_set {
+				fmt.Fprintf(o.out, "WARNING: Skipping DaemonSet-managed pod %v\n", pod.Name)
+				continue
+			}
+		}
+		_, found = pod.ObjectMeta.Annotations[types.ConfigMirrorAnnotationKey]
+		if found {
+			// Skip mirror pod
+			continue
+		}
+
+		pods = append(pods, pod)
+	}
+	return pods, err
+}
+
+// isDaemonSet takes a the value of the creator field, typically a serialized
+// SerializedReference, and decodes it.  It returns true iff the creator's kind
+// is DaemonSet.
+func (o *DrainOptions) isDaemonSet(creator string) (isDaemonSet bool, err error) {
+	isDaemonSet = false
+
+	r := o.factory.NewBuilder().
+		Stream(strings.NewReader(creator), "creator").
+		Do()
+	if err = r.Err(); err != nil {
+		return
+	}
+
+	err = r.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.Mapping.GroupVersionKind.Kind == "SerializedReference" && reflect.ValueOf(info.Object).Elem().FieldByName("Reference").FieldByName("Kind").String() == "DaemonSet" {
+			isDaemonSet = true
+		}
+		return nil
+	})
+
+	return
+}
+
+// scanForReplicationControllers verifies that all the pods have replication
+// controllers.  If any don't, we abort unless the user used --force.  In that
+// case we just warn.
+func (o *DrainOptions) scanForReplicationControllers(pods []api.Pod) error {
+	rcs, err := o.client.ReplicationControllers(api.NamespaceAll).List(nil, nil)
+	if err != nil {
+		return err
+	}
+	unreplicatedPodNames := []string{}
+
+	for _, pod := range pods {
+		replicated := false
+		for _, rc := range rcs.Items {
+			selector := labels.SelectorFromSet(rc.Spec.Selector)
+			if selector.Matches(labels.Set(pod.Labels)) {
+				replicated = true
+				break
+			}
+		}
+		if !replicated {
+			unreplicatedPodNames = append(unreplicatedPodNames, pod.Name)
+		}
+	}
+	if len(unreplicatedPodNames) > 0 {
+		joined := strings.Join(unreplicatedPodNames, ", ")
+		if !o.Force {
+			return fmt.Errorf("Refusing to continue due to unreplicated pods: %s (use --force to override)", joined)
+		}
+		fmt.Fprintf(o.out, "WARNING: Deleting these unreplicated pods: %s\n", joined)
```

Comment:
> scanForReplicationControllers is not actually deleting, just preparing for it.

The very next thing we do is send the delete call to the API server.  Are you concerned someone is scanning STDOUT and trying to draw some kind of strict timing conclusions from it?